src/adapter/odata/src/lib.rs

```diff
@@ -0,0 +1,390 @@
+    let ctx = ODataServiceContext::new(&host, &uri, catalog, account_name);
+    datafusion_odata::handlers::odata_metadata_handler(axum::Extension(Arc::new(ctx))).await
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub async fn odata_collection_handler_common(
+    axum::extract::Extension(catalog): axum::extract::Extension<Catalog>,
+    axum::extract::TypedHeader(host): axum::extract::TypedHeader<axum::headers::Host>,
+    axum::extract::OriginalUri(uri): axum::extract::OriginalUri,
+    account_name: Option<AccountName>,
+    dataset_name: DatasetName,
+    headers: axum::http::HeaderMap,
+    query: axum::extract::Query<QueryParamsRaw>,
+) -> axum::response::Response<String> {
+    let repo: Arc<dyn DatasetRepository> = catalog.get_one().unwrap();
+
+    let dataset_handle = match repo
+        .resolve_dataset_ref(&DatasetAlias::new(account_name, dataset_name).into())
+        .await
+    {
+        Ok(hdl) => Ok(hdl),
+        Err(GetDatasetError::NotFound(_)) => {
+            return axum::response::Response::builder()
+                .status(http::StatusCode::NOT_FOUND)
+                .body(Default::default())
+                .unwrap()
+        }
+        Err(e) => Err(e),
+    }
+    .unwrap();
+
+    let ctx = ODataCollectionContext::new(&host, &uri, catalog, dataset_handle);
+    datafusion_odata::handlers::odata_collection_handler(
+        axum::Extension(Arc::new(ctx)),
+        query,
+        headers,
+    )
+    .await
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Context
+///////////////////////////////////////////////////////////////////////////////
+
+struct ODataServiceContext {
+    catalog: Catalog,
+    account_name: Option<AccountName>,
+    service_base_url: String,
+}
+
+impl ODataServiceContext {
+    fn new(
+        host: &axum::headers::Host,
+        uri: &http::Uri,
+        catalog: Catalog,
+        account_name: Option<AccountName>,
+    ) -> Self {
+        // TODO: Find out scheme the API is served through (e.g. if there is an LB)
+        let scheme = "http";
+        let mut service_base_url = format!("{scheme}://{host}{uri}");
+        if service_base_url.ends_with('/') {
+            service_base_url.pop();
+        }
+
+        Self {
+            catalog,
+            account_name,
+            service_base_url,
+        }
+    }
+}
+
+// TODO: Authorization checks
+#[async_trait]
+impl ServiceContext for ODataServiceContext {
+    fn service_base_url(&self) -> String {
+        self.service_base_url.clone()
+    }
+
+    async fn list_collections(&self) -> Vec<Arc<dyn CollectionContext>> {
+        use futures::TryStreamExt;
+
+        let repo: Arc<dyn DatasetRepository> = self.catalog.get_one().unwrap();
+
+        let datasets = if let Some(account_name) = &self.account_name {
+            repo.get_datasets_by_owner(account_name)
+        } else {
+            repo.get_all_datasets()
+        };
```

Comment:
I don't like the idea of passing some default account name on this level. Perhaps we should change the `get_datasets_by_owner` API to support passing `None` in which case it would return datasets that belong to the currently authorized user...